File: aais-middleware/src/aais_tasks/graph_sync.ts
/**
 * Mythic: Graph ↔ AAIS task bridge
 * Engineering: syncFromGraph / syncToGraph — conflict detection + resolution policy
 */
import {
  callGraph,
  graphCreateTodoTask,
  graphListTodoTasks,
  type FetchLike,
} from "../provider_adapters/graph_client.js";
import type { AaisTask } from "./aais_task_model.js";
import type { AaisTaskStore } from "./aais_task_store.js";

export type SyncConflictPolicy = "prefer_aais" | "prefer_graph" | "report";

export interface SyncConflict {
  aaisTaskId: string;
  graphId: string;
  field: string;
  aaisValue: string;
  graphValue: string;
  resolution: "kept_aais" | "kept_graph" | "reported";
}

export interface GraphSyncResult {
  ok: boolean;
  reasonCode: string;
  imported?: AaisTask[];
  exported?: AaisTask[];
  conflicts?: SyncConflict[];
  error?: string;
  needsAuth?: boolean;
  policy?: SyncConflictPolicy;
}

function mapGraphStatus(s: unknown): AaisTask["status"] {
  const v = String(s || "").toLowerCase();
  if (v === "completed") return "completed";
  if (v === "inprogress" || v === "in_progress") return "inProgress";
  return "notStarted";
}

function statusToGraph(s: AaisTask["status"]): string {
  return s === "completed" ? "completed" : "notStarted";
}

function detectConflicts(
  aais: AaisTask,
  graphItem: Record<string, unknown>,
): { field: string; aaisValue: string; graphValue: string }[] {
  const diffs: { field: string; aaisValue: string; graphValue: string }[] = [];
  const gTitle = String(graphItem.title || "");
  if (gTitle && gTitle !== aais.title) {
    diffs.push({ field: "title", aaisValue: aais.title, graphValue: gTitle });
  }
  const gStatus = mapGraphStatus(graphItem.status);
  if (gStatus !== aais.status) {
    diffs.push({ field: "status", aaisValue: aais.status, graphValue: gStatus });
  }
  return diffs;
}

function resolvePolicy(explicit?: SyncConflictPolicy): SyncConflictPolicy {
  return (
    explicit ||
    (process.env.AAIS_GRAPH_SYNC_CONFLICT_POLICY as SyncConflictPolicy | undefined) ||
    "report"
  );
}

export async function syncFromGraph(
  store: AaisTaskStore,
  token: string | undefined,
  opts?: {
    fetchImpl?: FetchLike;
    listId?: string;
    conflictPolicy?: SyncConflictPolicy;
  },
): Promise<GraphSyncResult> {
  if (!token) {
    return {
      ok: false,
      needsAuth: true,
      reasonCode: "GRAPH_SYNC_NEEDS_AUTH",
      error: "Set AAIS_MS_GRAPH_TOKEN or connect Microsoft 365 for syncFromGraph.",
    };
  }
  const policy = resolvePolicy(opts?.conflictPolicy);
  const listed = await graphListTodoTasks(token, {
    fetchImpl: opts?.fetchImpl,
    listId: opts?.listId,
  });
  if (!listed.ok) {
    return {
      ok: false,
      reasonCode: listed.reasonCode,
      error: listed.error || "Graph list failed",
    };
  }
  const value = (listed.data as { value?: Record<string, unknown>[] })?.value || [];
  const imported: AaisTask[] = [];
  const conflicts: SyncConflict[] = [];

  for (const item of value) {
    const graphId = String(item.id || "");
    if (!graphId) continue;
    const existing = store.list().find((t) => t.graphId === graphId);
    if (existing) {
      const diffs = detectConflicts(existing, item);
      if (diffs.length === 0) {
        imported.push(existing);
        continue;
      }
      for (const d of diffs) {
        if (policy === "report") {
          conflicts.push({
            aaisTaskId: existing.id,
            graphId,
            ...d,
            resolution: "reported",
          });
          // no overwrite
        } else if (policy === "prefer_aais") {
          conflicts.push({
            aaisTaskId: existing.id,
            graphId,
            ...d,
            resolution: "kept_aais",
          });
        } else {
          // prefer_graph
          const patch: Record<string, unknown> = { source: "graph" };
          if (d.field === "title") patch.title = d.graphValue;
          if (d.field === "status") patch.status = mapGraphStatus(d.graphValue);
          const updated = store.update(existing.id, patch as never);
          if (updated) imported.push(updated);
          conflicts.push({
            aaisTaskId: existing.id,
            graphId,
            ...d,
            resolution: "kept_graph",
          });
        }
      }
      if (policy === "prefer_aais" || policy === "report") {
        imported.push(existing);
      }
      continue;
    }
    imported.push(
      store.create({
        title: String(item.title || "Graph task"),
        status: mapGraphStatus(item.status),
        dueDate: item.dueDateTime
          ? String((item.dueDateTime as { dateTime?: string }).dateTime || "")
          : undefined,
        source: "graph",
        graphId,
      }),
    );
  }
  return {
    ok: true,
    reasonCode: conflicts.some((c) => c.resolution === "reported")
      ? "GRAPH_SYNC_FROM_CONFLICTS"
      : "GRAPH_SYNC_FROM_OK",
    imported,
    conflicts,
    policy,
  };
}

export async function syncToGraph(
  store: AaisTaskStore,
  token: string | undefined,
  taskId: string,
  opts?: {
    fetchImpl?: FetchLike;
    listId?: string;
    conflictPolicy?: SyncConflictPolicy;
  },
): Promise<GraphSyncResult> {
  if (!token) {
    return {
      ok: false,
      needsAuth: true,
      reasonCode: "GRAPH_SYNC_NEEDS_AUTH",
      error: "Set AAIS_MS_GRAPH_TOKEN or connect Microsoft 365 for syncToGraph.",
    };
  }
  const policy = resolvePolicy(opts?.conflictPolicy);
  const task = store.get(taskId);
  if (!task) {
    return { ok: false, reasonCode: "AAIS_TASK_NOT_FOUND", error: `No task ${taskId}` };
  }
  const conflicts: SyncConflict[] = [];

  if (task.graphId) {
    const remote = await callGraph(
      token,
      `me/todo/lists/${encodeURIComponent(opts?.listId || "tasks")}/tasks/${encodeURIComponent(task.graphId)}`,
      "GET",
      undefined,
      { fetchImpl: opts?.fetchImpl },
    );
    if (remote.ok && remote.data && typeof remote.data === "object") {
      const diffs = detectConflicts(task, remote.data as Record<string, unknown>);
      for (const d of diffs) {
        if (policy === "report") {
          conflicts.push({
            aaisTaskId: task.id,
            graphId: task.graphId,
            ...d,
            resolution: "reported",
          });
        } else if (policy === "prefer_graph") {
          conflicts.push({
            aaisTaskId: task.id,
            graphId: task.graphId,
            ...d,
            resolution: "kept_graph",
          });
          const patch: Record<string, unknown> = { source: "graph" };
          if (d.field === "title") patch.title = d.graphValue;
          if (d.field === "status") patch.status = mapGraphStatus(d.graphValue);
          store.update(task.id, patch as never);
        } else {
          conflicts.push({
            aaisTaskId: task.id,
            graphId: task.graphId,
            ...d,
            resolution: "kept_aais",
          });
        }
      }
      if (policy === "report" && diffs.length > 0) {
        return {
          ok: true,
          reasonCode: "GRAPH_SYNC_TO_CONFLICTS",
          exported: [task],
          conflicts,
          policy,
        };
      }
      if (policy === "prefer_graph" && diffs.length > 0) {
        return {
          ok: true,
          reasonCode: "GRAPH_SYNC_TO_CONFLICTS",
          exported: [store.get(taskId) || task],
          conflicts,
          policy,
        };
      }
    }

    const patch = await callGraph(
      token,
      `me/todo/lists/${encodeURIComponent(opts?.listId || "tasks")}/tasks/${encodeURIComponent(task.graphId)}`,
      "PATCH",
      { title: task.title, status: statusToGraph(task.status) },
      { fetchImpl: opts?.fetchImpl },
    );
    if (!patch.ok) {
      return { ok: false, reasonCode: patch.reasonCode, error: patch.error, conflicts, policy };
    }
    return {
      ok: true,
      reasonCode: conflicts.length ? "GRAPH_SYNC_TO_CONFLICTS" : "GRAPH_SYNC_TO_OK",
      exported: [task],
      conflicts,
      policy,
    };
  }

  const created = await graphCreateTodoTask(token, task.title, {
    fetchImpl: opts?.fetchImpl,
    listId: opts?.listId,
  });
  if (!created.ok) {
    return { ok: false, reasonCode: created.reasonCode, error: created.error, policy };
  }
  const graphId = String((created.data as { id?: string })?.id || "");
  const updated = store.update(task.id, { graphId: graphId || undefined, source: "aais" });
  return {
    ok: true,
    reasonCode: "GRAPH_SYNC_TO_OK",
    exported: updated ? [updated] : [task],
    conflicts,
    policy,
  };
}

File: src/aais_tasks/graph_sync.py
"""Graph ↔ AAIS sync with conflict detection.

# Mythic: Graph ↔ AAIS task bridge
# Engineering: sync_from_graph / sync_to_graph — conflict policy
"""

from __future__ import annotations

import os
from typing import Any, Literal

from src.aais_tasks.aais_task_store import AaisTaskStore

SyncConflictPolicy = Literal["prefer_aais", "prefer_graph", "report"]


def _graph_client():
    # Lazy import avoids circular import via operator_middleware_plugs.__init__
    from src.operator_middleware_plugs.clients import graph_client as gc

    return gc


def _policy(explicit: str | None = None) -> SyncConflictPolicy:
    raw = (explicit or os.getenv("AAIS_GRAPH_SYNC_CONFLICT_POLICY") or "report").strip()
    if raw in {"prefer_aais", "prefer_graph", "report"}:
        return raw  # type: ignore[return-value]
    return "report"


def _map_status(s: Any) -> str:
    v = str(s or "").lower()
    if v == "completed":
        return "completed"
    if v in {"inprogress", "in_progress"}:
        return "inProgress"
    return "notStarted"


def _detect(aais: dict[str, Any], graph_item: dict[str, Any]) -> list[dict[str, str]]:
    diffs: list[dict[str, str]] = []
    g_title = str(graph_item.get("title") or "")
    if g_title and g_title != aais.get("title"):
        diffs.append({"field": "title", "aaisValue": str(aais.get("title")), "graphValue": g_title})
    g_status = _map_status(graph_item.get("status"))
    if g_status != aais.get("status"):
        diffs.append(
            {"field": "status", "aaisValue": str(aais.get("status")), "graphValue": g_status}
        )
    return diffs


def sync_from_graph(
    store: AaisTaskStore,
    token: str | None,
    *,
    list_id: str = "tasks",
    conflict_policy: str | None = None,
) -> dict[str, Any]:
    if not token:
        return {
            "ok": False,
            "needs_auth": True,
            "reason_code": "GRAPH_SYNC_NEEDS_AUTH",
            "error": "Set AAIS_MS_GRAPH_TOKEN for syncFromGraph.",
        }
    policy = _policy(conflict_policy)
    gc = _graph_client()
    listed = gc.graph_list_todo_tasks(token, list_id=list_id)
    if not listed.get("ok"):
        return {
            "ok": False,
            "reason_code": listed.get("reason_code"),
            "error": listed.get("error"),
        }
    value = (listed.get("data") or {}).get("value") if isinstance(listed.get("data"), dict) else []
    imported: list[dict[str, Any]] = []
    conflicts: list[dict[str, Any]] = []
    for item in value or []:
        if not isinstance(item, dict):
            continue
        graph_id = str(item.get("id") or "")
        if not graph_id:
            continue
        existing = next((t for t in store.list() if t.graph_id == graph_id), None)
        if existing:
            diffs = _detect(existing.to_dict(), item)
            if not diffs:
                imported.append(existing.to_dict())
                continue
            for d in diffs:
                if policy == "report":
                    conflicts.append(
                        {
                            "aaisTaskId": existing.id,
                            "graphId": graph_id,
                            **d,
                            "resolution": "reported",
                        }
                    )
                elif policy == "prefer_aais":
                    conflicts.append(
                        {
                            "aaisTaskId": existing.id,
                            "graphId": graph_id,
                            **d,
                            "resolution": "kept_aais",
                        }
                    )
                else:
                    patch: dict[str, Any] = {"source": "graph"}
                    if d["field"] == "title":
                        patch["title"] = d["graphValue"]
                    if d["field"] == "status":
                        patch["status"] = _map_status(d["graphValue"])
                    updated = store.update(existing.id, **patch)
                    if updated:
                        imported.append(updated.to_dict())
                    conflicts.append(
                        {
                            "aaisTaskId": existing.id,
                            "graphId": graph_id,
                            **d,
                            "resolution": "kept_graph",
                        }
                    )
            if policy in {"prefer_aais", "report"}:
                imported.append(existing.to_dict())
            continue
        task = store.create(
            title=str(item.get("title") or "Graph task"),
            status=_map_status(item.get("status")),
            source="graph",
            graph_id=graph_id,
        )
        imported.append(task.to_dict())
    return {
        "ok": True,
        "reason_code": (
            "GRAPH_SYNC_FROM_CONFLICTS"
            if any(c.get("resolution") == "reported" for c in conflicts)
            else "GRAPH_SYNC_FROM_OK"
        ),
        "imported": imported,
        "conflicts": conflicts,
        "policy": policy,
    }


def sync_to_graph(
    store: AaisTaskStore,
    token: str | None,
    task_id: str,
    *,
    list_id: str = "tasks",
    conflict_policy: str | None = None,
) -> dict[str, Any]:
    if not token:
        return {
            "ok": False,
            "needs_auth": True,
            "reason_code": "GRAPH_SYNC_NEEDS_AUTH",
            "error": "Set AAIS_MS_GRAPH_TOKEN for syncToGraph.",
        }
    policy = _policy(conflict_policy)
    task = store.get(task_id)
    if not task:
        return {"ok": False, "reason_code": "AAIS_TASK_NOT_FOUND", "error": f"No task {task_id}"}
    conflicts: list[dict[str, Any]] = []
    gc = _graph_client()
    if task.graph_id:
        remote = gc.call_graph(token, f"me/todo/lists/{list_id}/tasks/{task.graph_id}", method="GET")
        if remote.get("ok") and isinstance(remote.get("data"), dict):
            diffs = _detect(task.to_dict(), remote["data"])
            for d in diffs:
                if policy == "report":
                    conflicts.append(
                        {
                            "aaisTaskId": task.id,
                            "graphId": task.graph_id,
                            **d,
                            "resolution": "reported",
                        }
                    )
                elif policy == "prefer_graph":
                    patch = {"source": "graph"}
                    if d["field"] == "title":
                        patch["title"] = d["graphValue"]
                    if d["field"] == "status":
                        patch["status"] = _map_status(d["graphValue"])
                    store.update(task.id, **patch)
                    conflicts.append(
                        {
                            "aaisTaskId": task.id,
                            "graphId": task.graph_id,
                            **d,
                            "resolution": "kept_graph",
                        }
                    )
                else:
                    conflicts.append(
                        {
                            "aaisTaskId": task.id,
                            "graphId": task.graph_id,
                            **d,
                            "resolution": "kept_aais",
                        }
                    )
            if policy == "report" and diffs:
                return {
                    "ok": True,
                    "reason_code": "GRAPH_SYNC_TO_CONFLICTS",
                    "exported": [task.to_dict()],
                    "conflicts": conflicts,
                    "policy": policy,
                }
        patch = gc.call_graph(
            token,
            f"me/todo/lists/{list_id}/tasks/{task.graph_id}",
            method="PATCH",
            body={
                "title": task.title,
                "status": "completed" if task.status == "completed" else "notStarted",
            },
        )
        if not patch.get("ok"):
            return {
                "ok": False,
                "reason_code": patch.get("reason_code"),
                "error": patch.get("error"),
                "conflicts": conflicts,
                "policy": policy,
            }
        return {
            "ok": True,
            "reason_code": "GRAPH_SYNC_TO_CONFLICTS" if conflicts else "GRAPH_SYNC_TO_OK",
            "exported": [task.to_dict()],
            "conflicts": conflicts,
            "policy": policy,
        }
    created = gc.graph_create_todo_task(token, task.title, list_id=list_id)
    if not created.get("ok"):
        return {
            "ok": False,
            "reason_code": created.get("reason_code"),
            "error": created.get("error"),
            "policy": policy,
        }
    data = created.get("data") if isinstance(created.get("data"), dict) else {}
    graph_id = str((data or {}).get("id") or "")
    updated = store.update(task.id, graph_id=graph_id or None, source="aais")
    return {
        "ok": True,
        "reason_code": "GRAPH_SYNC_TO_OK",
        "exported": [(updated or task).to_dict()],
        "conflicts": conflicts,
        "policy": policy,
    }
